From 478d541d2fb425db014d17f65e2ce464f0414862 Mon Sep 17 00:00:00 2001
From: Ken Ma <make@marvell.com>
Date: Mon, 17 Jul 2017 17:49:56 +0800
Subject: [PATCH 1174/1345] ARM64: pinctrl: marvell: armada-37xx: correct
 south brigdge mpp23's group configuration

commit  174471025098a149decedae850e8aabc3eff9a1e from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Mpp23 in south bridge can be used for mii collision/mii tx err/gpio.
In South Bridge GPIO2 Selection register,
  - if bit14 is set, mpp23 is gpio;
  - if bit14 is clear, mpp23 is for mii:
	- if bit8 is set, mpp23 is for MII TX Error function(output);
	- if bit8 is clear, mpp23 is for MII Collision(input).
South bridge mpp23's gpio selection bitmask bit14 is different from
"rgmii" group pins(mpp6~mpp17) bitmask bit3, so it needs to be removed
from the group "rgmii"; otherwise when sd card uses mpp23 as vcc supply
gpio and requests gpio which causes "rgmii" group's selection bitmask
bit3 to be set, then mpp6~mpp17 will be gpio but not mii, this is not
we want, we only want mpp23 to be for sd card vcc supply gpio and want
to keep mpp6~mpp17 for rgmii still; this patch removes mpp23 from the
group "rgmii".
Since current Armada-37xx gpio driver only provides 2 functions for 1
group, yet mpp23 has three functions as above, so this patch follows
the way of ptp - father ptp group "ptp" is used to choose "ptp/mii" or
"gpio" and child ptp group such as "ptp_trig" is used to choose ptp
trigger generation function or mii carrier sense function. So mpp23 has
2 groups, the first group "mii" is used to select mpp23 as "mii" or
"gpio" while the second group "mii_col" is used to select pin as
"mii_col" or "mii_err" when the first group is selected as "mii".

Change-Id: I87781425d68cfd26fc21d00d62de7fda5e7e13dd
Reviewed-on: http://vgitil04.il.marvell.com:8080/41680
Tested-by: Wilson Ding <dingwei@marvell.com>
Verified-Armada37x0: Wilson Ding <dingwei@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Signed-off-by: Ken Ma <make@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/42777
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../pinctrl/marvell,armada-37xx-pinctrl.txt        |    6 +++++-
 drivers/pinctrl/mvebu/pinctrl-armada-37xx.c        |    4 ++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/Documentation/devicetree/bindings/pinctrl/marvell,armada-37xx-pinctrl.txt b/Documentation/devicetree/bindings/pinctrl/marvell,armada-37xx-pinctrl.txt
index e146985..df7c4ad 100644
--- a/Documentation/devicetree/bindings/pinctrl/marvell,armada-37xx-pinctrl.txt
+++ b/Documentation/devicetree/bindings/pinctrl/marvell,armada-37xx-pinctrl.txt
@@ -139,9 +139,13 @@ group ptp_trig
  - pin 58
  - functions ptp, mii
 
+group mii
+ - pin 59
+ - functions mii, gpio
+
 group mii_col
  - pin 59
- - functions mii, mii_err
+ - functions mii_col, mii_err
 
 GPIO subnode:
 
diff --git a/drivers/pinctrl/mvebu/pinctrl-armada-37xx.c b/drivers/pinctrl/mvebu/pinctrl-armada-37xx.c
index d8d2bf4..6b53a33 100644
--- a/drivers/pinctrl/mvebu/pinctrl-armada-37xx.c
+++ b/drivers/pinctrl/mvebu/pinctrl-armada-37xx.c
@@ -187,8 +187,8 @@ struct armada_37xx_pinctrl {
 	PIN_GRP_GPIO("ptp", 20, 3, BIT(11) | BIT(12) | BIT(13), "ptp"),
 	PIN_GRP("ptp_clk", 21, 1, BIT(6), "ptp", "mii"),
 	PIN_GRP("ptp_trig", 22, 1, BIT(7), "ptp", "mii"),
-	PIN_GRP_GPIO_3("mii_col", 23, 1, BIT(8) | BIT(14), 0, BIT(8), BIT(14),
-		       "mii", "mii_err"),
+	PIN_GRP_GPIO("mii", 23, 1, BIT(14), "mii"),
+	PIN_GRP("mii_col", 23, 1, BIT(8), "mii_col", "mii_err"),
 };
 
 const struct armada_37xx_pin_data armada_37xx_pin_nb = {
-- 
1.7.9.5

